aten/src/ATen/core/autocast/register_autocast_ops.cpp

```diff
@@ -0,0 +1,528 @@
+  AT_ERROR("torch.nn.functional.binary_cross_entropy and torch.nn.BCELoss are unsafe to autocast.\n"
+           "Many models use a sigmoid layer right before the binary cross entropy layer.\n"
+           "In this case, combine the two layers using torch.nn.functional.binary_cross_entropy_with_logits\n"
+           "or torch.nn.BCEWithLogitsLoss.  binary_cross_entropy_with_logits and BCEWithLogits are\n"
+           "safe to autocast.");
+}
+
+#ifndef USE_STATIC_DISPATCH
+namespace {
+/*****************************************************************************************************************
+This section performs load-time registration for autocast wrappers.
+
+It's debatable at which level operations should be patched.
+
+Option 1:  Patch only at the level of explicit calls into cudnn/cublas (cudnn_convolution, etc),
+because those are the code paths that are guaranteed to use Tensor Cores, therefore they're the ones that
+will benefit most from FP16.  One difficulty with this approach is that convolutions (and other ops) are wrapped
+in several layers of at::* calls.  Several non-explicitly-registered layers (e.g. at::convolution) could be called
+before we reach cudnn_convolution.  Because they are not explicitly registered, these layers would invoke the
+boxed fallback, which would exclude AutocastTensorId, so by the time at::cudnn_convolution is actually
+called, it wouldn't route back through the autocasting logic at all.
+
+Option 2:  Patch the Python-exposed surface of calls, to make 100% sure autograd history
+recording can't sneak in ahead of autocast.  This mirrors Apex most closely.
+
+I think Option 2 is the right answer for all ops, not just convolutions.  Option 2 is what I implement here.
+*****************************************************************************************************************/
+
+auto register_fallthrough = c10::Dispatcher::singleton()
+  .registerBackendFallbackKernel(DispatchKey::AutocastTensorId,
+                                 KernelFunction::makeFallthrough());
+
+// TODO:  Codegen the stuff below?  Ed said
+// > you are going to have to write the function definition at some point, I wouldn't try to get clever about it
+// Therefore, for the moment, this is all copy pasted in from VariableTypeEverything.cpp with appropriate substitutions.
+
+// Macros to reduce boilerplate somewhat.
+// Common case where registration signature matches redispatch signature
+// (that's why SIGNATURE is repeated in the WrapFunction instantiation)
+#define KERNEL(FUNC, REGISTER_SCHEMA, SIGNATURE, POLICY) \
+  .op(torch::RegisterOperators::options() \
+    .schema(REGISTER_SCHEMA) \
+    .kernel<SIGNATURE>(DispatchKey::AutocastTensorId, \
+    &WrapFunction<CastPolicy::POLICY, SIGNATURE, SIGNATURE, FUNC>::type::call) \
+    .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA))
+
+#define KERNEL_UNBOXED_ONLY(FUNC, REGISTER_SCHEMA, SIGNATURE, POLICY) \
```

Comment:
This is for all the function signatures that c10 dispatcher doesn't support boxing wrappers yet right? I'm a little surprised you didn't just make everything kernel unboxed only haha

Reply:
No, it's literally just because I'm imitating VariableType*.cpp's registration decisions (kernel vs kernelUnboxedOnly).  If registering everything as a kernelUnboxedOnly is more robust in some way, I can certainly do that.